src/Control/Effect/Interpret.hs

```diff
@@ -1,20 +1,74 @@
-{-# LANGUAGE GeneralizedNewtypeDeriving, MultiParamTypeClasses, RankNTypes, TypeOperators, UndecidableInstances #-}
+{-# language FlexibleContexts #-}
+{-# language FlexibleInstances #-}
+{-# language FunctionalDependencies #-}
+{-# language GeneralizedNewtypeDeriving #-}
+{-# language KindSignatures #-}
+{-# language PolyKinds #-}
+{-# language RankNTypes #-}
+{-# language ScopedTypeVariables #-}
+{-# language TypeApplications #-}
+{-# language TypeOperators #-}
+{-# language UndecidableInstances #-}
+
 module Control.Effect.Interpret
 ( runInterpret
-, InterpretC(..)
 , runInterpretState
-, InterpretStateC(..)
+, InterpretC(..)
 ) where
 
 import Control.Applicative (Alternative(..))
 import Control.Effect.Carrier
-import Control.Effect.Reader
 import Control.Effect.State
 import Control.Monad (MonadPlus(..))
 import Control.Monad.Fail
 import Control.Monad.Fix
 import Control.Monad.IO.Class
 import Control.Monad.Trans.Class
+import Unsafe.Coerce ( unsafeCoerce )
+
+
+newtype InterpretC ( s :: * ) ( sig :: ( * -> * ) -> * -> * ) m a =
+  InterpretC { unInterpretC :: m a }
+  deriving
+    (Functor, Applicative, Monad, MonadFix, MonadFail, MonadIO, MonadPlus, Alternative)
+
+
+instance MonadTrans ( InterpretC s sig ) where
+  lift =
+    InterpretC
+
+
+newtype Handler sig m =
+  Handler { runHandler :: forall s x. sig ( InterpretC s sig m ) x -> InterpretC s sig m x }
+
+
+newtype Tagged a b =
+  Tagged { unTag :: b }
+
+
+class Reifies ( s :: * ) a | s -> a where
+  reflect :: Tagged s a
+
+
+data Skolem
+
+
+newtype Magic a r =
```

Comment:
Could you please document the declarations used in the definition here? Even if they’re not exported, a quick note describing the purpose of these pieces would be very helpful, even if it’s ultimately “magical” (in the sense that it’s reliant on implementation details of ghc’s calling conventions w.r.t. single-method dictionaries).